File: soundkit-opus/tools/raw_pcm_roundtrip.c
#include <errno.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include <opus.h>

#define CHANNELS 2
#define SAMPLE_RATE 48000
#define MAX_PACKET_BYTES 1275

static void usage(void) {
    fprintf(stderr,
            "usage: raw_pcm_roundtrip_c [--pcm-bits 16|24] <frame-size> "
            "<bitrate> <cbr|vbr> <input.raw> <output.f32le>\n");
    exit(2);
}

static int parse_positive_int(const char *value) {
    char *end = NULL;
    long parsed = strtol(value, &end, 10);
    if (end == value || *end != '\0' || parsed <= 0 || parsed > INT32_MAX) {
        usage();
    }
    return (int)parsed;
}

static int valid_frame_size(int frame_size) {
    return frame_size == 120 || frame_size == 240 || frame_size == 480 ||
           frame_size == 960;
}

static void *read_samples(
    const char *path,
    size_t sample_size,
    size_t *sample_count,
    const char *format
) {
    FILE *file = fopen(path, "rb");
    if (!file) {
        fprintf(stderr, "cannot open %s: %s\n", path, strerror(errno));
        exit(1);
    }
    if (fseek(file, 0, SEEK_END) != 0) {
        perror("fseek");
        exit(1);
    }
    long byte_count = ftell(file);
    if (byte_count < 0 || byte_count % (long)sample_size != 0) {
        fprintf(stderr, "invalid %s input length\n", format);
        exit(1);
    }
    rewind(file);
    *sample_count = (size_t)byte_count / sample_size;
    void *samples = malloc(*sample_count * sample_size);
    if (!samples) {
        fprintf(stderr, "out of memory\n");
        exit(1);
    }
    if (fread(samples, sample_size, *sample_count, file) != *sample_count) {
        fprintf(stderr, "cannot read %s\n", path);
        exit(1);
    }
    if (fclose(file) != 0) {
        perror("fclose");
        exit(1);
    }
    return samples;
}

static void write_samples(const char *path, const float *samples, size_t sample_count) {
    FILE *file = fopen(path, "wb");
    if (!file) {
        fprintf(stderr, "cannot create %s: %s\n", path, strerror(errno));
        exit(1);
    }
    if (fwrite(samples, sizeof(float), sample_count, file) != sample_count) {
        fprintf(stderr, "cannot write %s\n", path);
        exit(1);
    }
    if (fclose(file) != 0) {
        perror("fclose");
        exit(1);
    }
}

static void check_opus(int result, const char *operation) {
    if (result != OPUS_OK) {
        fprintf(stderr, "%s failed: %s\n", operation, opus_strerror(result));
        exit(1);
    }
}

int main(int argc, char **argv) {
    int pcm_bits = 0;
    int offset = 1;
    if (argc >= 3 && strcmp(argv[1], "--pcm-bits") == 0) {
        pcm_bits = parse_positive_int(argv[2]);
        if (pcm_bits != 16 && pcm_bits != 24) {
            usage();
        }
        offset = 3;
    }
    if (argc != offset + 5) {
        usage();
    }
    int frame_size = parse_positive_int(argv[offset]);
    int bitrate = parse_positive_int(argv[offset + 1]);
    if (!valid_frame_size(frame_size)) {
        usage();
    }
    int vbr;
    if (strcmp(argv[offset + 2], "cbr") == 0) {
        vbr = 0;
    } else if (strcmp(argv[offset + 2], "vbr") == 0) {
        vbr = 1;
    } else {
        usage();
    }

    size_t sample_count = 0;
    float *input = NULL;
    opus_int16 *input_i16 = NULL;
    opus_int32 *input_i24 = NULL;
    if (pcm_bits == 16) {
        input_i16 = (opus_int16 *)read_samples(
            argv[offset + 3], sizeof(*input_i16), &sample_count, "s16le");
    } else if (pcm_bits == 24) {
        input_i24 = (opus_int32 *)read_samples(
            argv[offset + 3], sizeof(*input_i24), &sample_count, "s32le");
        for (size_t sample = 0; sample < sample_count; sample++) {
            if (input_i24[sample] < -8388608 || input_i24[sample] > 8388607) {
                fprintf(stderr, "input contains a sample outside signed 24-bit range\n");
                return 1;
            }
        }
    } else {
        input = (float *)read_samples(
            argv[offset + 3], sizeof(*input), &sample_count, "f32le");
    }
    if (sample_count % CHANNELS != 0) {
        fprintf(stderr, "input has a partial stereo frame\n");
        return 1;
    }
    float *output = (float *)calloc(sample_count + (size_t)frame_size * CHANNELS,
                                    sizeof(float));
    float *padded = pcm_bits == 0
        ? (float *)calloc((size_t)frame_size * CHANNELS, sizeof(float))
        : NULL;
    float *decoded = pcm_bits == 0
        ? (float *)malloc((size_t)frame_size * CHANNELS * sizeof(float))
        : NULL;
    opus_int16 *padded_i16 = pcm_bits == 16
        ? (opus_int16 *)calloc((size_t)frame_size * CHANNELS, sizeof(*padded_i16))
        : NULL;
    opus_int16 *decoded_i16 = pcm_bits == 16
        ? (opus_int16 *)malloc((size_t)frame_size * CHANNELS * sizeof(*decoded_i16))
        : NULL;
    opus_int32 *padded_i24 = pcm_bits == 24
        ? (opus_int32 *)calloc((size_t)frame_size * CHANNELS, sizeof(*padded_i24))
        : NULL;
    opus_int32 *decoded_i24 = pcm_bits == 24
        ? (opus_int32 *)malloc((size_t)frame_size * CHANNELS * sizeof(*decoded_i24))
        : NULL;
    unsigned char packet[MAX_PACKET_BYTES];
    if (!output || (pcm_bits == 0 && (!padded || !decoded)) ||
        (pcm_bits == 16 && (!padded_i16 || !decoded_i16)) ||
        (pcm_bits == 24 && (!padded_i24 || !decoded_i24))) {
        fprintf(stderr, "out of memory\n");
        return 1;
    }

    int error = OPUS_OK;
    OpusEncoder *encoder = opus_encoder_create(
        SAMPLE_RATE, CHANNELS, OPUS_APPLICATION_AUDIO, &error);
    check_opus(error, "opus_encoder_create");
    OpusDecoder *decoder = opus_decoder_create(SAMPLE_RATE, CHANNELS, &error);
    check_opus(error, "opus_decoder_create");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_BITRATE(bitrate)),
               "OPUS_SET_BITRATE");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_VBR(vbr)), "OPUS_SET_VBR");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_VBR_CONSTRAINT(vbr)),
               "OPUS_SET_VBR_CONSTRAINT");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_COMPLEXITY(10)),
               "OPUS_SET_COMPLEXITY");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_BANDWIDTH(OPUS_BANDWIDTH_FULLBAND)),
               "OPUS_SET_BANDWIDTH");
    check_opus(opus_encoder_ctl(encoder, OPUS_SET_SIGNAL(OPUS_SIGNAL_MUSIC)),
               "OPUS_SET_SIGNAL");

    size_t frame_samples = (size_t)frame_size * CHANNELS;
    size_t output_offset = 0;
    size_t packet_bytes = 0;
    size_t packet_count = 0;
    int packet_min = MAX_PACKET_BYTES;
    int packet_max = 0;
    for (size_t offset = 0; offset < sample_count; offset += frame_samples) {
        size_t remaining = sample_count - offset;
        size_t copied = remaining < frame_samples ? remaining : frame_samples;
        int packet_len;
        if (pcm_bits == 16) {
            memset(padded_i16, 0, frame_samples * sizeof(*padded_i16));
            memcpy(padded_i16, input_i16 + offset, copied * sizeof(*padded_i16));
            packet_len = opus_encode(
                encoder, padded_i16, frame_size, packet, MAX_PACKET_BYTES);
        } else if (pcm_bits == 24) {
            memset(padded_i24, 0, frame_samples * sizeof(*padded_i24));
            memcpy(padded_i24, input_i24 + offset, copied * sizeof(*padded_i24));
            packet_len = (int)opus_encode24(
                encoder, padded_i24, frame_size, packet, MAX_PACKET_BYTES);
        } else {
            memset(padded, 0, frame_samples * sizeof(*padded));
            memcpy(padded, input + offset, copied * sizeof(*padded));
            packet_len = opus_encode_float(
                encoder, padded, frame_size, packet, MAX_PACKET_BYTES);
        }
        if (packet_len < 0) {
            fprintf(stderr, "opus_encode_float failed: %s\n", opus_strerror(packet_len));
            return 1;
        }
        int decoded_frames;
        if (pcm_bits == 16) {
            decoded_frames = opus_decode(
                decoder, packet, packet_len, decoded_i16, frame_size, 0);
        } else if (pcm_bits == 24) {
            decoded_frames = opus_decode24(
                decoder, packet, packet_len, decoded_i24, frame_size, 0);
        } else {
            decoded_frames = opus_decode_float(
                decoder, packet, packet_len, decoded, frame_size, 0);
        }
        if (decoded_frames != frame_size) {
            fprintf(stderr, "opus_decode_float returned %d frames\n", decoded_frames);
            return 1;
        }
        if (pcm_bits == 16) {
            for (size_t sample = 0; sample < frame_samples; sample++) {
                output[output_offset + sample] =
                    (float)decoded_i16[sample] * (1.0f / 32768.0f);
            }
        } else if (pcm_bits == 24) {
            for (size_t sample = 0; sample < frame_samples; sample++) {
                output[output_offset + sample] =
                    (float)decoded_i24[sample] * (1.0f / 8388608.0f);
            }
        } else {
            memcpy(output + output_offset, decoded, frame_samples * sizeof(float));
        }
        output_offset += frame_samples;
        packet_bytes += (size_t)packet_len;
        packet_count++;
        if (packet_len < packet_min) packet_min = packet_len;
        if (packet_len > packet_max) packet_max = packet_len;
    }
    write_samples(argv[offset + 4], output, sample_count);

    const char *pcm_label = pcm_bits == 16 ? "16" : pcm_bits == 24 ? "24" : "float";
    printf("{\"codec\":\"%s\",\"sample_rate\":%d,\"channels\":%d,\"pcm_bits\":\"%s\","
           "\"frame_size\":%d,\"bitrate\":%d,\"mode\":\"%s\","
           "\"packets\":%zu,\"packet_bytes\":%zu,\"packet_min\":%d,"
           "\"packet_max\":%d}\n",
           opus_get_version_string(), SAMPLE_RATE, CHANNELS, pcm_label, frame_size, bitrate,
           vbr ? "vbr" : "cbr", packet_count, packet_bytes, packet_min, packet_max);

    opus_encoder_destroy(encoder);
    opus_decoder_destroy(decoder);
    free(input);
    free(input_i16);
    free(input_i24);
    free(output);
    free(padded);
    free(decoded);
    free(padded_i16);
    free(decoded_i16);
    free(padded_i24);
    free(decoded_i24);
    return 0;
}

File: soundkit-opus/examples/raw_pcm_roundtrip.rs
use soundkit_opus::{Application, Decoder, Encoder};
use std::env;
use std::fs;
use std::io;
use std::path::Path;

const CHANNELS: usize = 2;
const SAMPLE_RATE: i32 = 48_000;
const VALID_FRAME_SIZES: [usize; 4] = [120, 240, 480, 960];

fn usage() -> ! {
    eprintln!(
        "usage: raw_pcm_roundtrip [--pcm-bits 16|24] <frame-size> <bitrate> \
         <cbr|vbr> <input.raw> <output.f32le>"
    );
    std::process::exit(2);
}

enum PcmInput {
    F32(Vec<f32>),
    I16(Vec<i16>),
    I24(Vec<i32>),
}

impl PcmInput {
    fn len(&self) -> usize {
        match self {
            Self::F32(samples) => samples.len(),
            Self::I16(samples) => samples.len(),
            Self::I24(samples) => samples.len(),
        }
    }

    fn label(&self) -> &'static str {
        match self {
            Self::F32(_) => "float",
            Self::I16(_) => "16",
            Self::I24(_) => "24",
        }
    }
}

fn read_f32le(path: &Path) -> io::Result<Vec<f32>> {
    let data = fs::read(path)?;
    if data.len() % 4 != 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input has a partial f32 sample",
        ));
    }
    let samples = data
        .chunks_exact(4)
        .map(|bytes| f32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
        .collect::<Vec<_>>();
    if samples.iter().any(|sample| !sample.is_finite()) {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input contains a non-finite sample",
        ));
    }
    Ok(samples)
}

fn read_i16le(path: &Path) -> io::Result<Vec<i16>> {
    let data = fs::read(path)?;
    if data.len() % 2 != 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input has a partial i16 sample",
        ));
    }
    Ok(data
        .chunks_exact(2)
        .map(|bytes| i16::from_le_bytes([bytes[0], bytes[1]]))
        .collect())
}

fn read_i24_s32le(path: &Path) -> io::Result<Vec<i32>> {
    let data = fs::read(path)?;
    if data.len() % 4 != 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input has a partial i32 sample",
        ));
    }
    let samples = data
        .chunks_exact(4)
        .map(|bytes| i32::from_le_bytes([bytes[0], bytes[1], bytes[2], bytes[3]]))
        .collect::<Vec<_>>();
    if samples
        .iter()
        .any(|sample| !(-8_388_608..=8_388_607).contains(sample))
    {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input contains a sample outside signed 24-bit range",
        ));
    }
    Ok(samples)
}

fn write_f32le(path: &Path, samples: &[f32]) -> io::Result<()> {
    let mut data = Vec::with_capacity(samples.len() * 4);
    for sample in samples {
        data.extend_from_slice(&sample.to_le_bytes());
    }
    fs::write(path, data)
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = env::args().collect::<Vec<_>>();
    let (pcm_bits, offset) = if args.get(1).map(String::as_str) == Some("--pcm-bits") {
        let bits = args
            .get(2)
            .and_then(|value| value.parse::<u8>().ok())
            .filter(|value| matches!(value, 16 | 24))
            .unwrap_or_else(|| usage());
        (Some(bits), 3)
    } else {
        (None, 1)
    };
    if args.len() != offset + 5 {
        usage();
    }
    let frame_size = args[offset].parse::<usize>().unwrap_or_else(|_| usage());
    if !VALID_FRAME_SIZES.contains(&frame_size) {
        usage();
    }
    let bitrate = args[offset + 1].parse::<i32>().unwrap_or_else(|_| usage());
    let vbr = match args[offset + 2].as_str() {
        "cbr" => false,
        "vbr" => true,
        _ => usage(),
    };
    let input_path = Path::new(&args[offset + 3]);
    let output_path = Path::new(&args[offset + 4]);
    let input = match pcm_bits {
        Some(16) => PcmInput::I16(read_i16le(input_path)?),
        Some(24) => PcmInput::I24(read_i24_s32le(input_path)?),
        _ => PcmInput::F32(read_f32le(input_path)?),
    };
    if input.len() % CHANNELS != 0 {
        return Err(io::Error::new(
            io::ErrorKind::InvalidData,
            "input has a partial stereo frame",
        )
        .into());
    }

    let total_samples = input.len();
    let frame_samples = frame_size * CHANNELS;
    let mut encoder = Encoder::new(SAMPLE_RATE, CHANNELS, Application::Audio)?;
    encoder.set_bitrate(bitrate)?;
    encoder.set_vbr(vbr)?;
    let mut decoder = Decoder::new(SAMPLE_RATE, CHANNELS)?;
    let mut output = Vec::with_capacity(total_samples + frame_samples);
    let mut packet_bytes = 0usize;
    let mut packet_count = 0usize;
    let mut packet_min = usize::MAX;
    let mut packet_max = 0usize;

    for offset in (0..total_samples).step_by(frame_samples) {
        let copied = (total_samples - offset).min(frame_samples);
        let packet = match &input {
            PcmInput::F32(samples) => {
                let mut padded = vec![0.0f32; frame_samples];
                padded[..copied].copy_from_slice(&samples[offset..offset + copied]);
                encoder.encode_f32(&padded, frame_size)?
            }
            PcmInput::I16(samples) => {
                let mut padded = vec![0i16; frame_samples];
                padded[..copied].copy_from_slice(&samples[offset..offset + copied]);
                encoder.encode_i16(&padded, frame_size)?
            }
            PcmInput::I24(samples) => {
                let mut padded = vec![0i32; frame_samples];
                padded[..copied].copy_from_slice(&samples[offset..offset + copied]);
                encoder.encode_i24(&padded, frame_size)?
            }
        };
        packet_bytes += packet.len();
        packet_count += 1;
        packet_min = packet_min.min(packet.len());
        packet_max = packet_max.max(packet.len());
        match &input {
            PcmInput::F32(_) => output.extend(decoder.decode_f32(&packet, false)?),
            PcmInput::I16(_) => output.extend(
                decoder
                    .decode_i16(&packet, false)?
                    .into_iter()
                    .map(|sample| f32::from(sample) * (1.0 / 32_768.0)),
            ),
            PcmInput::I24(_) => output.extend(
                decoder
                    .decode_i24(&packet, false)?
                    .into_iter()
                    .map(|sample| sample as f32 * (1.0 / 8_388_608.0)),
            ),
        }
    }
    output.truncate(total_samples);
    write_f32le(output_path, &output)?;

    let mode = if vbr { "vbr" } else { "cbr" };
    println!(
        "{{\"codec\":\"soundkit-opus\",\"sample_rate\":{SAMPLE_RATE},\"channels\":{CHANNELS},\"pcm_bits\":\"{}\",\"frame_size\":{frame_size},\"bitrate\":{bitrate},\"mode\":\"{mode}\",\"packets\":{packet_count},\"packet_bytes\":{packet_bytes},\"packet_min\":{packet_min},\"packet_max\":{packet_max}}}",
        input.label()
    );
    Ok(())
}
